### DISCUSSION TITLE
User-scaling with bounds relaxation

### DISCUSSION THREAD

**Discussion by bknueven** [coin-or/Ipopt#555] [Q&A] [answered]
### Summary:
When using `nlp_scaling_method=user-scaling`, the variable bounds are relaxed by Eqn. (35) in the implementation paper and then scaled according to the provided factor. It would friendlier to some models I'm working with (and I think more natural) to do things in the reverse order: apply the variable scaling factor to the bounds, and then relax the bounds according to Eqn. (35).

I'm wondering if the current behavior is as-intended, in which case I will work around it, or if this is not as-intended and is a matter of a (hopefully) straightforward patch.

### Details:
I'm attempting to use scaling factors to scale some variables in a large-scale NLP. I've reduced the issue down to a trivial linear program, expressed in Pyomo:
```python
import pyomo.environ as pyo
m = pyo.ConcreteModel()
factor = 1.0e-12
m.x = pyo.Var(bounds=(0.5*factor, 1.5*factor), initialize=factor)
m.scaling_factor = pyo.Suffix(direction=pyo.Suffix.EXPORT)
m.scaling_factor[m.x] = 1./factor
 
m.o = pyo.Objective(expr=m.x/factor)

ipopt = pyo.SolverFactory("ipopt")
ipopt.options["print_level"] = 8
ipopt.options["nlp_scaling_method"] = "user-scaling"
ipopt.solve(m, tee=True, symbolic_solver_labels=True)

print(f"optimal value: {m.x.value}")
```
Or in math:
```
minimize 1.0e+12*x  subject to x within [5.0e-13, 1.5e-12] 
```
Notice that I'm scaling the variable x by `1.0e+12` to attempt to bring its value close to `1.0`, and utilizing the `user-scaling` `nlp_scaling_method`.

The optimal solution reported by Ipopt:
```
optimal value: -9.999499999997493e-09
```
which is well outside, in relative terms, to the original lower bound of `5.0e-13`.

Looking at the log for Ipopt, what's apparently happening is that Eqn. (35) in the implementation paper is applied to the *unscaled* variable bounds, which are then scaled. Below is some sample output:
```
Ipopt 3.14.4: print_level=8
nlp_scaling_method=user-scaling


List of options:

                                    Name   Value                # times used
                      nlp_scaling_method = user-scaling              1
                             print_level = 8                         2

******************************************************************************
This program contains Ipopt, a library for large-scale nonlinear optimization.
 Ipopt is released as open source code under the Eclipse Public License (EPL).
         For more information visit https://github.com/coin-or/Ipopt
******************************************************************************

This is Ipopt version 3.14.4, running with linear solver MUMPS 5.2.1.

Number of nonzeros in equality constraint Jacobian...:        0
Number of nonzeros in inequality constraint Jacobian.:        0
Number of nonzeros in Lagrangian Hessian.............:        0

objective scaling factor = 1
x scaling provided
No c scaling provided
No d scaling provided
DenseVector "x scaling vector" with 1 elements:
x scaling vector[    1]{x}= 1.0000000000000000e+12
DenseVector "modified x_L scaled" with 1 elements:
modified x_L scaled[    1]{x}=-9.9995000000000000e+03
DenseVector "modified x_U scaled" with 1 elements:
modified x_U scaled[    1]{x}= 1.0001500000000000e+04
DenseVector "modified d_L scaled" with 0 elements:
DenseVector "modified d_U scaled" with 0 elements:
DenseVector "initial x unscaled" with 1 elements:
initial x unscaled[    1]{x}= 9.9999999999999998e-13
Initial values of x sufficiently inside the bounds.
Initial values of s sufficiently inside the bounds.
Total number of variables............................:        1
                     variables with only lower bounds:        0
                variables with lower and upper bounds:        1
                     variables with only upper bounds:        0
Total number of equality constraints.................:        0
Total number of inequality constraints...............:        0
        inequality constraints with only lower bounds:        0
   inequality constraints with lower and upper bounds:        0
        inequality constraints with only upper bounds:        0
```

For this problem, it would be better to *scale* the variable bounds and *then* apply Eqn. (35) from the implementation paper. Doing this by-hand (setting `bound_relax_factor=0` and applying Eqn. (35) manually to the model), I converge to an solution equivalent to being well-scaled in the first place: `4.999999925092976e-13`.

**Comment by svigerske** [comment, marked as answer]
I cannot say whether things were intentional or not, but I presume that to change this, one would touch `OrigIpoptNLP::InitializeStructures()`. This is where the scaling is computed and the bounds are relaxed:

https://github.com/coin-or/Ipopt/blob/a62679dd2cd292a9e3aa64ad05df37f5adfdece5/src/Algorithm/IpOrigIpoptNLP.cpp#L231-L348

To avoid too many changes, a possible way would be to compute the amount of bound relaxation on the scaled bounds and then scale them back to the original problem and apply them to the original bounds. There should be an option to choose between computing the bound relaxation on the original and the scaled bounds.
